### ISSUE TITLE
understand and apply sky visplane hashing optimization

### ISSUE THREAD

**Issue by fabiangreffrath** [fabiangreffrath/woof#1107]
That is, find out how and why this apparently works:
https://github.com/JNechaevsky/inter-doom/commit/a2407f93262bb48eb0e8c86965a554ce44627cba

**Comment by fabiangreffrath** [comment]
@JNechaevsky Can you see the same effect in PrBoom+/DSDA-Doom or is this restricted to "our" family of ports?

**Comment by JNechaevsky** [comment]
Sure, gladly. I'll post my result with/without that correction, just need some extra time to compile Pr+ and DSDA. Plainsphere-2 with full view from south part of the map should be ideal for benchmarking.

**Comment by JNechaevsky** [comment]
**DSDA** (software, 640x480, no vsync)

<details>

WIth `lightlevel = 0`
![image](https://github.com/fabiangreffrath/woof/assets/21193394/4cf438fb-7bc1-4613-b023-017c5bd96fff)

Without `lightlevel = 0`
![image](https://github.com/fabiangreffrath/woof/assets/21193394/fadebb47-1555-4086-ac32-3cd0a1af7111)

</details>

**PrBoom+** (software, 640x480, no vsync)

<details>

With `lightlevel = 0`
![image](https://github.com/fabiangreffrath/woof/assets/21193394/2e2b7908-432a-4620-8fb3-0ce59644a0c9)

Without `lightlevel = 0`
![image](https://github.com/fabiangreffrath/woof/assets/21193394/086b7e45-dd3e-4f7f-93da-171f8500de4e)

</details>

To summarize: for FPS it is matters no if `lightlevel = 0` is here or not, it's stable in both cases. Amount of rendered visplanes, hovewer, is different, w/o `lightlevel = 0` engine render less of them. 🤔

**Comment by JNechaevsky** [comment]
One more run for Woof, now on Sunder MAP15:

<details>

WIth `lightlevel = 0`
![image](https://github.com/fabiangreffrath/woof/assets/21193394/e9ed9cf4-24f2-4854-9334-0600dc3bf2fd)

Without `lightlevel = 0`
![image](https://github.com/fabiangreffrath/woof/assets/21193394/e53f21de-efd0-45d9-955e-65ca71b60b52)

</details>

**Comment by fabiangreffrath** [comment]
I think, if the game renders *faster* with *more* visplanes, then our process of extending the visplanes' dimensions is too slow. This happens in `R_CheckPlane()` and the one difference between us and PrBoom+ is that we use `insigned int` for `pl->top[]` whereas they use `unsigned short`.

**Comment by JNechaevsky** [comment]
You mean this line? https://github.com/fabiangreffrath/woof/blob/master/src/r_plane.c#L302
Replacing `0xffffffffu` with `SHRT_MAX` resulting a notable fps drop-off in Woof.

**Comment by fabiangreffrath** [comment]
Yes, because it's wrong! You'd have to change the type of the `top` struct element in `visplane_t` to `unsigned short` as well.

**Comment by fabiangreffrath** [comment]
And there is also line 459.

**Comment by fabiangreffrath** [comment]
And we don't check for `SHRT_MAX` then but `0xffffu`.

**Comment by JNechaevsky** [comment]
Just tried by your recommendations. Hey, it gave about ~7-10 extra fps on Sunder, in the scene from my screenshot from above! 😧

**Comment by fabiangreffrath** [comment]
Wow, I didn't expect that! The difference between the two cases with and without `lightlevel = 0` was only 5 FPS, so this even over-compensated.

You once made a test-WAD when we first discussed this issue without literally thousands of different floor planes under a sky. It was called `vplanes.wad` or similar, do you remember?

**Comment by JNechaevsky** [comment]
I've lost it... But take a try this one, it's far more extreme and shows about ~5000 visplanes in starting view: [pln.zip](https://github.com/fabiangreffrath/woof/files/11745225/pln.zip)

**Comment by JNechaevsky** [comment]
Two more things: replacing [this](https://github.com/fabiangreffrath/woof/blob/master/src/r_plane.c#L230) and [this](https://github.com/fabiangreffrath/woof/blob/master/src/r_plane.c#L270) by using code from DSDA-Doom ([here](https://github.com/kraflab/dsda-doom/blob/master/prboom2/src/r_plane.c#LL284C1-L285C35) and [here](https://github.com/kraflab/dsda-doom/blob/master/prboom2/src/r_plane.c#LL284C1-L285C35), except we are using `0xffffu` instead of `SHRT_MAX`) gives about 10-15 extra fps on Sunder, at least on my Corei7-12700F. That's increabile. 😮

**Comment by rfomin** [comment]
> Two more things: replacing [this](https://github.com/fabiangreffrath/woof/blob/master/src/r_plane.c#L230) and [this](https://github.com/fabiangreffrath/woof/blob/master/src/r_plane.c#L270) by using code from DSDA-Doom ([here](https://github.com/kraflab/dsda-doom/blob/master/prboom2/src/r_plane.c#LL284C1-L285C35) and [here](https://github.com/kraflab/dsda-doom/blob/master/prboom2/src/r_plane.c#LL284C1-L285C35), except we are using 0xffffu instead of SHRT_MAX) gives about 10-15 extra fps on Sunder, at least on my Corei7-12700F. That's increabile.

I tried it on an old Core i5 and didn't get any difference. Do you use release builds? Otherwise, you will end up with an unoptimized `memset` function. I used this code:
```c
{
    int i;
    for (i = 0; i != sizeof(new_pl->top); i++)
      new_pl->top[i] = USHRT_MAX;
}
```
Although maybe `SCREENWIDTH` is enough, instead of `MAX_SCREENWIDTH` (`sizeof(new_pl->top)`).

Anyway, I think we should use `unsigned short` if we can.

**Comment by JNechaevsky** [comment]
Sure, I'm using: `cmake -G "Ninja" -D CMAKE_BUILD_TYPE="Release" -S . -B build`. You can simplify such inner loop to:
```
    for (int i = 0 ; i != viewwidth ; i++)
    {
        check->top[i] = 0xffffu;
    }
```
No sure, though, is `viewwidth` correct here... Also, such trick with `int i = ...` inside `for` loop may not work in all complilers, Watcom C for sure. 😀

Could you please take a try this savegame from Sunder to compare before/after? FPS will not be super stable, but at least it is possible to catch by eye the difference. [woofsav0.zip](https://github.com/fabiangreffrath/woof/files/11750663/woofsav0.zip)

**Comment by rfomin** [comment]
> Could you please take a try this savegame from Sunder to compare before/after? FPS will not be super stable, but at least it is possible to catch by eye the difference. [woofsav0.zip](https://github.com/fabiangreffrath/woof/files/11750663/woofsav0.zip)

Still no difference for me. I think "catch by eye" not accurate enough, it's better to record a demo, run it several times and compare avg/max/min FPS.

> Also, such trick with `int i = ...` inside `for` loop may not work in all complilers, Watcom C for sure. 😀

Watcom C supports the C89 standard, `int i` inside a `for` loop is a C99 innovation. Although we use C99, we still support C89 style code for consistency with the rest of the codebase. In any case, this is pure "syntax sugar" and should not affect performance.

**Comment by JNechaevsky** [comment]
Good idea, thanks! Take a try this demo: [pln-demo.zip](https://github.com/fabiangreffrath/woof/files/11753152/pln-demo.zip) with my demo wad from above ([pln.zip](https://github.com/fabiangreffrath/woof/files/11745225/pln.zip)). It's about ~30 seconds of starring to 5000 visplanes. My -timedemo results of average fps on windowed mode 640x480 w/o screen wiping, five runs for for both approaches:

With `memset` approach: 
* 105, 103, 104, 105, 102

With `for` approach:
* 127, 127, 130, 131, 131

**Comment by rfomin** [comment]
I get less FPS with the `for` approach, although not by much. Maybe you have some changes elsewhere? Please do a PR or post the diff here.
Another theory is compiler differences - I use MSVC.

**Comment by fabiangreffrath** [comment]
I cannot confirm your finding wrt `for()` vs `memset()`. Quite the contrary, I get a massive fps penalty (no actual numbers yet).

Two ideas why `memset()` may perform badly on your system: 
The sheet number of bytes to be set. Could you please try replacing `sizeof(pl->top)` with `viewwidth * sizeof(*pl->top)`?
Unaligned access. Could you please turn the `pad1` element of `struct visplane_t` into an array of two (or more) shorts, so that the relative address of the `top` element inside the struct is a multiple of 4?

**Comment by JNechaevsky** [comment]
What the... I'm using just simple replacements in actual Woof code (i.e. with latest changes of visplanes to short types):
```
R_DupPlane
      new_pl->minx = start;
      new_pl->maxx = stop;
---   memset(new_pl->top, 0xff, sizeof new_pl->top);
+++   for (int i = 0 ; i != viewwidth ; i++)
+++     new_pl->top[i] = 0xffffu;
```
```
R_FindPlane
check->xoffs = xoffs;               // killough 2/28/98: Save offsets
  check->yoffs = yoffs;

---  memset (check->top, 0xff, sizeof check->top);
+++  for (int i = 0 ; i != viewwidth ; i++)
+++    check->top[i] = 0xffffu;
```
And it's GCC (gcc.exe (Rev6, Built by MSYS2 project) 13.1.0).

**Comment by rfomin** [comment]
My diff:
<details>

```diff
diff --git a/src/r_defs.h b/src/r_defs.h
index 796c08c..ece1b83 100644
--- a/src/r_defs.h
+++ b/src/r_defs.h
@@ -438,11 +438,11 @@ typedef struct visplane
   fixed_t height;
   fixed_t xoffs, yoffs;         // killough 2/28/98: Support scrolling flats
   // [FG] 32-bit integer math
-  unsigned int pad1;          // leave pads for [minx-1]/[maxx+1]
-  unsigned int top[MAX_SCREENWIDTH];
-  unsigned int pad2, pad3;    // killough 2/8/98, 4/25/98
-  unsigned int bottom[MAX_SCREENWIDTH];
-  unsigned int pad4;
+  unsigned short pad1;          // leave pads for [minx-1]/[maxx+1]
+  unsigned short top[MAX_SCREENWIDTH];
+  unsigned short pad2, pad3;    // killough 2/8/98, 4/25/98
+  unsigned short bottom[MAX_SCREENWIDTH];
+  unsigned short pad4;
 } visplane_t;
 
 #endif
diff --git a/src/r_plane.c b/src/r_plane.c
index 73fec13..8a452d1 100644
--- a/src/r_plane.c
+++ b/src/r_plane.c
@@ -237,7 +237,12 @@ visplane_t *R_DupPlane(const visplane_t *pl, int start, int stop)
       new_pl->yoffs = pl->yoffs;
       new_pl->minx = start;
       new_pl->maxx = stop;
-      memset(new_pl->top, 0xff, sizeof new_pl->top);
+      //memset(new_pl->top, 0xff, sizeof new_pl->top);
+      {
+        int i;
+        for (i = 0; i < viewwidth; i++)
+          new_pl->top[i] = USHRT_MAX;
+      }
 
       return new_pl;
 }
@@ -277,7 +282,12 @@ visplane_t *R_FindPlane(fixed_t height, int picnum, int lightlevel,
   check->xoffs = xoffs;               // killough 2/28/98: Save offsets
   check->yoffs = yoffs;
 
-  memset (check->top, 0xff, sizeof check->top);
+  //memset (check->top, 0xff, sizeof check->top);
+  {
+    int i;
+    for (i = 0; i < viewwidth; i++)
+      check->top[i] = USHRT_MAX;
+  }
 
   return check;
 }
@@ -299,7 +309,7 @@ visplane_t *R_CheckPlane(visplane_t *pl, int start, int stop)
   else
     unionh  = pl->maxx, intrh  = stop;
 
-  for (x=intrl ; x <= intrh && pl->top[x] == 0xffffffffu; x++) // [FG] 32-bit integer math
+  for (x=intrl ; x <= intrh && pl->top[x] == USHRT_MAX; x++) // [FG] 32-bit integer math
     ;
 
   if (x > intrh)
@@ -456,7 +466,7 @@ static void do_draw_plane(visplane_t *pl)
 
         stop = pl->maxx + 1;
         planezlight = zlight[light];
-        pl->top[pl->minx-1] = pl->top[stop] = 0xffffffffu; // [FG] 32-bit integer math
+        pl->top[pl->minx-1] = pl->top[stop] = USHRT_MAX; // [FG] 32-bit integer math
 
         for (x = pl->minx ; x <= stop ; x++)
           R_MakeSpans(x,pl->top[x-1],pl->bottom[x-1],pl->top[x],pl->bottom[x]);
```
</details>

**Comment by fabiangreffrath** [comment]
IIRC, SDL even has a `SDL_memset4()` function which is optimized to write 4 bytes at once. Maybe we should benchmark this as well.

**Comment by fabiangreffrath** [comment]
https://github.com/fabiangreffrath/woof/commit/fcd26e812a79640142e5e99917993bb553edc682
